From faf604bbf07004459e3473033698039f622e829e Mon Sep 17 00:00:00 2001
From: Olivier Fourdan <ofourdan@redhat.com>
Date: Fri, 31 Jul 2026 16:20:56 +0200
Subject: [PATCH libXi 3/3] size_classes/copy_classes: enforce XI2 per-type
 class payload sizes

Bounding any_wire->length against the reply/event end pointer is not
enough: a class length that fits in the buffer can still be shorter
than the type-specific wire struct, and button/key counts can claim
more bytes than that class contains. copy_classes() then reads past
the payload when converting XI_DeviceChanged events (and when parsing
XIQueryDevice replies).

Validate per-type minimum sizes and that button/key payloads fit in
the declared class length, and reject DeviceChanged allocations that
would overflow size_t.

Found by AISLE in partnership with Red Hat

AISLE-Report-PSIRTSUPT-13717

CVE-2026-93543

v2: Use size_t in size_classes and reject length overflow.
v3: XI2 keycodes are 4 bytes (Peter Hutterer)

Reported-by: Aisle Research
Signed-off-by: Olivier Fourdan <ofourdan@redhat.com>
Assisted-by: AI
Part-of: <https://gitlab.freedesktop.org/xorg/lib/libxi/-/merge_requests/23>
(cherry picked from commit f885619bb67f1ddec5d4e40ff3eff1a3269a0069)
---
 src/XExtInt.c | 101 +++++++++++++++++++++++++++++++++++++++++++++++---
 1 file changed, 95 insertions(+), 6 deletions(-)

diff --git a/src/XExtInt.c b/src/XExtInt.c
index 2842c44..4278622 100644
--- a/src/XExtInt.c
+++ b/src/XExtInt.c
@@ -56,6 +56,7 @@ SOFTWARE.
 
 #include <stdio.h>
 #include <stdint.h>
+#include <limits.h>
 #include <X11/extensions/XI.h>
 #include <X11/extensions/XI2.h>
 #include <X11/extensions/XIproto.h>
@@ -1674,15 +1675,93 @@ validate_class_header(char *ptr, char *end, xXIAnyInfo **any_out)
     return 0;
 }
 
+/*
+ * Reject class records whose type-specific header or payload cannot fit in
+ * the declared wire length (length is in 4-byte units).
+ */
+static int
+validate_class_payload(xXIAnyInfo *any)
+{
+    size_t cls_len = (size_t) any->length * 4;
+
+    switch (any->type) {
+    case XIButtonClass:
+        {
+            xXIButtonInfo *b = (xXIButtonInfo *) any;
+            size_t mask_size, labels_size, need;
+
+            if (cls_len < sizeof(xXIButtonInfo))
+                return -1;
+
+            mask_size = ((size_t) b->num_buttons + 7) / 8;
+            mask_size = (mask_size + 3) / 4 * 4;
+            if (b->num_buttons >
+                (SIZE_MAX - sizeof(xXIButtonInfo) - mask_size) / 4)
+                return -1;
+
+            labels_size = (size_t) b->num_buttons * 4;
+            need = sizeof(xXIButtonInfo) + mask_size + labels_size;
+            if (need > cls_len)
+                return -1;
+
+            break;
+        }
+    case XIKeyClass:
+        {
+            xXIKeyInfo *k = (xXIKeyInfo *) any;
+
+            if (cls_len < sizeof(xXIKeyInfo))
+                return -1;
+
+            /* copy_classes reads num_keycodes bytes from the wire */
+            if ((size_t) k->num_keycodes > (cls_len - sizeof(xXIKeyInfo)) / 4)
+                return -1;
+
+            break;
+        }
+    case XIValuatorClass:
+        if (cls_len < sizeof(xXIValuatorInfo))
+            return -1;
+        break;
+    case XIScrollClass:
+        if (cls_len < sizeof(xXIScrollInfo))
+            return -1;
+        break;
+    case XITouchClass:
+        if (cls_len < sizeof(xXITouchInfo))
+            return -1;
+        break;
+    case XIGestureClass:
+        if (cls_len < sizeof(xXIGestureInfo))
+            return -1;
+        break;
+    default:
+        break;
+    }
+
+    return 0;
+}
+
 _X_HIDDEN int
 size_classes(xXIAnyInfo* from, int nclasses, char *end)
 {
-    int len, i;
+    size_t len;
+    size_t ptr_bytes;
+    int i;
     xXIAnyInfo *any_wire;
     char *ptr_wire;
 
+    if (nclasses < 0)
+        return -1;
+    if ((size_t) nclasses > SIZE_MAX / sizeof(XIAnyClassInfo *))
+        return -1;
+
+    ptr_bytes = (size_t) nclasses * sizeof(XIAnyClassInfo *);
+    if (ptr_bytes > (size_t) INT_MAX)
+        return -1;
+
     /* len for to->classes */
-    len = pad_to_double(nclasses * sizeof(XIAnyClassInfo*));
+    len = (size_t) pad_to_double((int) ptr_bytes);
     ptr_wire = (char*)from;
     for (i = 0; i < nclasses; i++)
     {
@@ -1691,6 +1770,9 @@ size_classes(xXIAnyInfo* from, int nclasses, char *end)
         if (validate_class_header(ptr_wire, end, &any_wire) < 0)
             return -1;
 
+        if (validate_class_payload(any_wire) < 0)
+            return -1;
+
         switch(any_wire->type)
         {
             case XIButtonClass:
@@ -1715,11 +1797,15 @@ size_classes(xXIAnyInfo* from, int nclasses, char *end)
                 break;
         }
 
-        len += l;
+        if (l < 0 || len > SIZE_MAX - (size_t) l)
+            return -1;
+        len += (size_t) l;
         ptr_wire += any_wire->length * 4;
     }
 
-    return len;
+    if (len > (size_t) INT_MAX)
+        return -1;
+    return (int) len;
 }
 
 #define FP3232_TO_DOUBLE(x) ((double) (x).integral + (double) (x).frac / (1ULL << 32))
@@ -1757,6 +1843,9 @@ copy_classes(XIDeviceInfo* to, xXIAnyInfo* from, int *nclasses, char *end)
         if (validate_class_header(ptr_wire, end, &any_wire) < 0)
             return -1;
 
+        if (validate_class_payload(any_wire) < 0)
+            return -1;
+
         switch(any_wire->type)
         {
             case XIButtonClass:
@@ -1911,10 +2000,10 @@ wireToDeviceChangedEvent(xXIDeviceChangedEvent *in, XGenericEventCookie *cookie)
     XIDeviceInfo info;
     int len;
     int nclasses = in->num_classes;
-    char *end = (char *)in + sizeof(xEvent) + in->length * 4;
+    char *end = (char *)in + sizeof(xXIDeviceChangedEvent) + in->length * 4;
 
     len = size_classes((xXIAnyInfo*)&in[1], in->num_classes, end);
-    if (len < 0)
+    if (len < 0 || (size_t)len > SIZE_MAX - sizeof(XIDeviceChangedEvent))
         return 0;
 
     cookie->data = out = malloc(sizeof(XIDeviceChangedEvent) + len);
-- 
2.56.0

